### DISCUSSION TITLE
KEDA Trigger from Prometheus to External Microservice

### DISCUSSION THREAD

**Discussion by Edward250250** [kedacore/keda#1545] [Q&A / Need Help] [answered]
Hello All,

I have a question regarding using external triggers. Before I am using prometheus trigger to fetch the customer metrics and then scale. 

I can set directly the query to prometheus and filter on my deployment :

apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: prometheus-scaledobject
  namespace: default
spec:
  scaleTargetRef:
    name: my-deployment
  pollingInterval: 15
  cooldownPeriod: 15
  minReplicaCount: 2
  maxReplicaCount: 20
  **triggers:
  - type: prometheus
    metadata:
      serverAddress: http://<prometheus-host>:9090
      metricName: my_custom_metrics_name
      threshold: '90'
      query: "sum(session_active_status{namespace=\"{{ .Values.keda.namespace }}\",service=\"{{ template "service.fullname" . }}\"}) * 100"**

However, now I have the need to add another microservice between Prometheus and KEDA to add some info.
Instead of KEDA fetch metrics from Prometheus directly, now it talks to my microservice and my microservice talk to Prometheus.
Kind of like a proxy. So I checked the KEDA docs and it looks I need to use the External Trigger :

triggers:
- type: external
  metadata:
    scalerAddress: external-scaler-service:8080
    tlsCertFile : /path/to/tls/cert.pem

I am not sure how can I migrate from my prometheus trigger to my external microservice as a trigger.
Can you please kindly help here, or point me to some docs ? I was trying to find the yaml schema for KEDA which contains all the fields I could use, but no luck. I am guess I could just move the same fields under metadata from Prometheus to External trigger and it should work ? For example the same query if my microservices return all the metrics in json format ? I am not sure. 

Can you please help, thank you in advance : )

**Comment by zroubalik** [comment, marked as answer]
In order to use `external` scaler, you need to provide the implementation of the scaler by yourself. You need to implement gRPC contract, see the doc: https://keda.sh/docs/2.1/concepts/external-scalers/

**Reply by Edward250250** [reply]
Thank you very much for your reply, I followed this page and created my js gRPC server and test gRPC client just to call getMetrics().

This maybe a silly question but what does this getEarthquakeCount() do in the javascript part? https://keda.sh/docs/2.1/concepts/external-scalers/#6-implementing-getmetrics
It looks like calling this method to get the final return value to keda hpa.

I guess I cannot use async/await in this gRPC function so it is doing in the call back. 
However, when I try to call this method from my test client, it tells me this getEarthquakeCount is not defined.
This makes me wondering if I am doing it correctly by creating a simple test client to call this method...
I made some changes and defined a new getEarthquakeCount() function just to return a same number to test.
But my client seems hangs there waiting for server response and I guess if it is in real scenario, the keda client is actually running it in a loop, so keep pulling the data and calling this getMetrics() method. 

...
      **getEarthquakeCount**((err, count) => {
        if (err) {
          callback({
            code: grpc.status.INTERNAL,
            details: err,
          })
        } else {
          callback(null, {
            metricValues: [{
              metricName: 'earthquakeThreshold',
              metricValue: count,
            }]
          })
        }

Thank you for your time and reply, I am trying to get better understanding about this part : )

**Reply by zroubalik** [reply]
Not sure if I fully understand your problem, for in that particular example, you'll need to implement `getEarthquakeCount()`, you can see the actual code in here: https://keda.sh/docs/2.1/concepts/external-scalers/#3-implementing-isactive instead of returning a bool, you should return the count.

**Reply by Edward250250** [reply]
Thank you. I created a function for that : )

**Reply by zroubalik** [reply]
If you have found anything confusing in the external scaler docs, would be nice if you could open a PR with a fix, to help us improve that experience :)
